File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/wordy/wordy.js
//
// This is only a SKELETON file for the 'Wordy' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const answer = (question) => {
  if (!/^What is .+\?$/.test(question)) {
    throw new Error('Unknown operation');
  }

  const mathPart = question.slice(8, -1).trim();

  if (mathPart === '') {
    throw new Error('Syntax error');
  }

  const tokens = mathPart.match(/-?\d+|multiplied by|divided by|plus|minus/g);

  // Check if all input was consumed by valid tokens
  if (!tokens || tokens.length === 0 || tokens.join(' ') !== mathPart) {
    throw new Error('Unknown operation');
  }

  if (!/^-?\d+$/.test(tokens[0])) {
    throw new Error('Syntax error');
  }

  if (tokens.length % 2 === 0) {
    throw new Error('Syntax error');
  }

  for (let i = 1; i < tokens.length; i += 2) {
    const operator = tokens[i];
    const numberToken = tokens[i + 1];

    if (!['plus', 'minus', 'multiplied by', 'divided by'].includes(operator)) {
      throw new Error('Unknown operation');
    }

    if (!/^-?\d+$/.test(numberToken)) {
      throw new Error('Syntax error');
    }
  }

  let result = parseInt(tokens[0], 10);

  for (let i = 1; i < tokens.length; i += 2) {
    const operator = tokens[i];
    const num = parseInt(tokens[i + 1], 10);

    switch (operator) {
      case 'plus':
        result += num;
        break;
      case 'minus':
        result -= num;
        break;
      case 'multiplied by':
        result *= num;
        break;
      case 'divided by':
        if (num === 0) {
          throw new Error('Division by zero');
        }
        if (result % num !== 0) {
          throw new Error('Non-integer division');
        }
        result /= num;
        break;
      default:
        throw new Error('Unknown operation');
    }
  }

  return result;
};

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/wordy/wordy.go
package wordy

import (
	"strconv"
	"strings"
)

func Answer(question string) (int, bool) {
	if !strings.HasPrefix(question, "What is ") || !strings.HasSuffix(question, "?") {
		return 0, false
	}

	content := strings.TrimSpace(question[8 : len(question)-1])
	if content == "" {
		return 0, false
	}

	words := strings.Fields(content)
	var tokens []string

	for i := 0; i < len(words); {
		word := words[i]
		if word == "multiplied" && i+1 < len(words) && words[i+1] == "by" {
			tokens = append(tokens, "multiplied by")
			i += 2
		} else if word == "divided" && i+1 < len(words) && words[i+1] == "by" {
			tokens = append(tokens, "divided by")
			i += 2
		} else if word == "plus" || word == "minus" {
			tokens = append(tokens, word)
			i++
		} else {
			tokens = append(tokens, word)
			i++
		}
	}

	if len(tokens) == 0 {
		return 0, false
	}

	num, err := strconv.Atoi(tokens[0])
	if err != nil {
		return 0, false
	}

	result := num

	for i := 1; i < len(tokens); i += 2 {
		if i+1 >= len(tokens) {
			return 0, false
		}

		op := tokens[i]
		nextNum, err := strconv.Atoi(tokens[i+1])
		if err != nil {
			return 0, false
		}

		switch op {
		case "plus":
			result += nextNum
		case "minus":
			result -= nextNum
		case "multiplied by":
			result *= nextNum
		case "divided by":
			if nextNum == 0 {
				return 0, false
			}
			result /= nextNum
		default:
			return 0, false
		}
	}

	return result, true
}
